client/history/client.go

```diff
@@ -1020,6 +1015,122 @@ func (c *clientImpl) NotifyFailoverMarkers(
 	return nil
 }
 
+func (c *clientImpl) GetCrossClusterTasks(
+	ctx context.Context,
+	request *types.GetCrossClusterTasksRequest,
+	opts ...yarpc.CallOption,
+) (*types.GetCrossClusterTasksResponse, error) {
+	requestByClient := make(map[Client]*types.GetCrossClusterTasksRequest)
+	for _, shardID := range request.GetShardIDs() {
+		client, err := c.getClientForShardID(int(shardID))
+		if err != nil {
+			return nil, err
+		}
+
+		if _, ok := requestByClient[client]; !ok {
+			requestByClient[client] = &types.GetCrossClusterTasksRequest{
+				TargetCluster: request.TargetCluster,
+			}
+		}
+		requestByClient[client].ShardIDs = append(requestByClient[client].ShardIDs, shardID)
+	}
+
+	var wg sync.WaitGroup
+	futureByClient := make(map[Client]future.Future, len(requestByClient))
+	for client, req := range requestByClient {
+		future, settable := future.NewFuture()
+		wg.Add(1)
+		go func() {
+			defer wg.Done()
+			settable.Set(client.GetCrossClusterTasks(ctx, req))
+		}()
+
+		futureByClient[client] = future
+	}
+
+	// preserve 5% timeout to return partial of the result if context is timing out
+	ctx, cancel := c.createChildContext(ctx, 0.05)
+	defer cancel()
+	response := &types.GetCrossClusterTasksResponse{
+		TasksByShard: make(map[int32][]*types.CrossClusterTaskRequest),
+	}
+	var err error
+	for _, future := range futureByClient {
+		var resp *types.GetCrossClusterTasksResponse
+		if futureErr := future.Get(ctx, &resp); futureErr != nil {
+			c.logger.Error("Failed to get cross cluster tasks", tag.Error(futureErr))
+			// TODO: return error for each shard and perform backoff at shard level.
+			// and ensure every shardID in request has a response (either tasks or failed cause).
+			//
+			// for _, failedShardID := range requestByClient[client].ShardIDs {
+			// 	response.FailedCauseByShard[failedShardID] = ...
+			// }
+			//
+			// for now following the pattern for getting replication tasks:
+			// ignore errors other than service busy, so that task fetcher in target
+			// cluster can slow down.
+			if common.IsServiceBusyError(futureErr) {
+				err = futureErr
+			}
+		} else {
+			for shardID, tasks := range resp.TasksByShard {
+				response.TasksByShard[shardID] = tasks
+			}
+		}
+	}
+	wg.Wait()
```

Comment:
should this before future get?

Reply:
Both will work I think?

I guess I can probably just remove the `wg`, once all futures are unblocked those goroutines will eventually be completed.

Reply:
Updated. Removed the wg usage and move the logic for reducing ctx timeout before creating the futures.